src/main/java/org/folio/service/orders/PurchaseOrderLineService.java

```diff
@@ -130,22 +131,54 @@ public Future<Void> saveOrderLine(CompositePoLine compositePoLine, List<Location
 
-  public Future<Void> saveOrderLines(PoLineCollection poLineCollection, RequestContext requestContext) {
-    RequestEntry requestEntry = new RequestEntry(BATCH_ENDPOINT);
-    return collectResultsOnSuccess(transform(poLineCollection.getPoLines(), poLine -> updateSearchLocations(poLine, requestContext)))
-      .compose(v -> restClient.put(requestEntry, poLineCollection, requestContext));
+  public Future<Void> saveOrderLinesWithoutSearchLocationsUpdate(List<PoLine> orderLines, RequestContext requestContext) {
+    List<PoLineCollection> poLineCollections = getPartitionedPoLines(orderLines);
+    return saveOrderLinesCollections(poLineCollections, requestContext);
   }
 
+  public Future<Void> saveOrderLinesWithLocations(List<PoLineLocationsPair> pairs, RequestContext requestContext) {
+    List<PoLineCollection> poLineCollections = getPartitionedPoLines(pairs.stream().map(PoLineLocationsPair::getPoLine).toList());
+
+    for (PoLineCollection collection: poLineCollections) {
+      for (PoLine poLine: collection.getPoLines()) {
+        List<Location> locations = pairs.stream()
+          .filter(pair -> StringUtils.equals(pair.getPoLine().getId(), poLine.getId()))
+          .findFirst().orElseThrow(() -> new NoSuchElementException("No matching PoLine found"))
+          .getLocations();
+        updateSearchLocations(poLine, locations, requestContext);
+      }
+    }
+
+    return saveOrderLinesCollections(poLineCollections, requestContext);
+  }
 
   public Future<Void> saveOrderLines(List<PoLine> orderLines, RequestContext requestContext) {
-    List<PoLineCollection> poLineCollections = ListUtils.partition(orderLines, PO_LINE_BATCH_PARTITION_SIZE)
+    List<PoLineCollection> poLineCollections = getPartitionedPoLines(orderLines);
+
+    for (PoLineCollection collection: poLineCollections) {
+      for (PoLine poLine: collection.getPoLines()) {
+        updateSearchLocations(poLine, requestContext);
+      }
+    }
+
+    return saveOrderLinesCollections(poLineCollections, requestContext);
+  }
+
+  private List<PoLineCollection> getPartitionedPoLines(List<PoLine> orderLines) {
+    return ListUtils.partition(orderLines, PO_LINE_BATCH_PARTITION_SIZE)
       .stream()
       .map(lines -> new PoLineCollection().withPoLines(lines).withTotalRecords(lines.size()))
       .toList();
+  }
 
-    return GenericCompositeFuture.join(poLineCollections.stream()
-      .map(poLineCollection -> saveOrderLines(poLineCollection, requestContext))
-      .toList())
+  private Future<Void> saveOrderLinesCollections(List<PoLineCollection> poLineCollections, RequestContext requestContext) {
+    return GenericCompositeFuture.join(poLineCollections
+      .stream()
+      .map(poLineCollection -> {
+        logger.info("saveOrderLines:: start saving {} po lines in batch", poLineCollection.getTotalRecords());
+        RequestEntry requestEntry = new RequestEntry(BATCH_ENDPOINT);
+        return restClient.put(requestEntry, poLineCollection, requestContext);
```

Comment:
I think we can put poLineCollections to one poLineCollections to avoid multiple requests by creating some method before sending request
```
       List<PoLine> combinedPoLines = poLIneCollections.stream().flatMap(collections -> collections.getPolines.stream())
       .toList() 
        int totalRecords = combinedPoLines.size();
        return new PoLineCollections(combinedPoLines, totalRecords);
```

Reply:
it was an intent to make partitioning here to update each 100 po lines at a time, see method getPartitionedPoLines()

Reply:
Ok, I see